assembler produces off-by-two branch displacement

ceharris opened the pull request:
Below is a very simple assembly routine that counts down using the Y register.
```
: branch-bug  ( n -- )
    [
        0 ldy.zx
    -->
        dey
        <b bne
        inx
        inx
    ]
    ;
```

Using latest `master-64tass`, the routine assembles as follows.

```
nt: 800  xt: 80F  header: 01 0A 65 C0 07
flags: HC 0 NN 0 AN 0 IM 0 CO 0 DC 0 LC 0 FP 1 | UF 0 ST 0
size (decimal): 7

080F  B4 00 88 D0 FF E8 E8                              .......

80F      0 ldy.zx
811        dey
812     FF bne      813 ^
814        inx
815        inx
```

Note that the displacement here is $ff (-1). Given that PC will be $814 when this displacement is added, the effective address for the branch is the second byte of the `bne` instruction itself at $813. Whoops. The correct displacement is $fd (-3) = $814 - 3 = $811. I'm guessing that maybe the `<b` pseudo-instruction isn't accounting for the size of the `BNE` instruction itself.

Looking at the source for `assembler.asm`, it seems the code is attempting to account for the size of the branch instruction, but it neglects put the result of the two decrement instructions onto the data stack with `sta 0,x` before returning.

```
xt_asm_back_branch:
                ; We arrive here with ( addr-l ) of the label on the stack and
                ; then subtract the current address
                jsr w_here             ; ( addr-l addr-h )
                jsr w_minus            ; ( offset )

                ; We subtract two more because of the branch instruction itself
                dea
                dea
                
z_asm_back_branch:
                rts
```

This PR adds an instruction to save the resulting displacement at TOS.

ceharris committed:
assembler branch displacement off-by-two

SamCoVT commented:
This definitely looks like a bug.

Looking at w_minus (the word run just before the decrements), it touches the MSB last (as expected for handling borrowing), so A has the MSB in it after that word and not the LSB.

We probably shouldn't be relying on side effects of the `MINUS` (A is considered available between Forth words), so this should probably be changed to 
```
        dec 0,x
        dec 0,x
```
to directly do the decrementing on the Forth data stack. 

`<b` is a relatively new word, and I don't think it has any tests.  It might also be worth writing a test for it in `/tests/asm.fs`.  The tests usually create a predictable result on the data stack and then check for that result.  They looks like:
`T{ do stuff with results on stack -> result1 result2 }T`
`T{` doesn't actually do anything.  `->` saves the current values on the stack and clears the stack.  `}T` compares the saved results to what is on the stack now, printing an error if they differ in values or number of items on the data stack.  A passing test is silent.

You can run the tests with one of the following commands:
`make ctests` - if you have a C compiler installed.  This uses the c65 simulator included with Tali and is by far the fastest method (seconds).
`make ptests` - only works on Linux (maybe OSX) and uses multiple instances of py65mon in parallel.  Will hog your CPU for a minute or two.
`make tests` - run tests in py65mon sequantially - if python and py65mon are installed, this works on pretty much all platforms, but it takes minutes to complete.

You don't need to include the tests results in your PR - I'll run the tests and update the results myself when I merge.

If writing a test is something you are interested in, great!  If not, just let me know and I'm happy to write one for it.  One helpful word to know about (in the assembler-wordlist) for testing assembly is `PUSH-A` which pushes the value in A onto the Forth data stack (you can see `PUSH-A` being tested [here](https://github.com/SamCoVT/TaliForth2/blob/0509ed5cc669cfbf55c445ed5984acfc21cf3618/tests/asm.fs#L69)).  The test could loop and do something with A and then push the result for comparison to an expected value.

ceharris committed:
don't rely on w_minus internals

ceharris committed:
add simple test case for <B

ceharris committed:
minor cleanup

ceharris commented:
Ironically, in the simple demo I made, the correct displacement happens to be $FD and since the MSB of the subtraction is $FF, decrementing A twice appeared to work correctly. I didn't even look at `w_minus`, since I assumed whoever wrote the two `dea` statements knew the state of A after `w_minus`.  😁 

Changed from decrementing A to decrementing the LSB of TOS as suggested.
Added a very simple test case, mostly for the experience of doing it. More robust test cases would actually look at the generated displacement, but at least this one crashes/fails if the displacement is wrong.

SamCoVT commented:
I think that looks reasonable.  Your simple test is significantly better than the (non-existent) previous test we were using.  The tester can usually detect if Tali crashes, but won't catch an infinite loop.
Is this PR all set for merging?

ceharris commented:
All set to merge. Thanks!
